react/features/video-quality/subscriber.js

```diff
@@ -289,110 +415,77 @@ function _updateReceiverVideoConstraints({ getState }) {
-                receiverConstraints.onStageSources = [ largeVideoSourceName ];
-            }
+    // Tile view.
+    if (shouldDisplayTileView(state)) {
+        if (!visibleRemoteTrackSourceNames?.length) {
+            return;
         }
 
+        visibleRemoteTrackSourceNames.forEach(sourceName => {
+            receiverConstraints.constraints[sourceName] = { 'maxHeight': maxFrameHeightForTileView };
+        });
+
+        // Prioritize screenshare in tile view.
         if (remoteScreenSharesSourceNames?.length) {
-            remoteScreenSharesSourceNames.forEach(sourceName => {
-                receiverConstraints.constraints[sourceName] = { 'maxHeight': VIDEO_QUALITY_LEVELS.ULTRA };
-            });
+            receiverConstraints[sourceNameSignaling ? 'selectedSources' : 'selectedEndpoints']
+                = remoteScreenSharesSourceNames;
         }
 
+    // Stage view.
     } else {
-        receiverConstraints = {
-            constraints: {},
-            defaultConstraints: { 'maxHeight': VIDEO_QUALITY_LEVELS.NONE },
-            lastN,
-            onStageEndpoints: [],
-            selectedEndpoints: []
-        };
-
-        // Tile view.
-        if (shouldDisplayTileView(state)) {
-            if (!visibleRemoteParticipants?.size) {
-                return;
-            }
+        if (!visibleRemoteTrackSourceNames?.length && !largeVideoSourceName && !activeParticipantsSources?.length) {
+            return;
+        }
 
-            visibleRemoteParticipants.forEach(participantId => {
-                receiverConstraints.constraints[participantId] = { 'maxHeight': maxFrameHeight };
+        if (visibleRemoteTrackSourceNames?.length) {
+            visibleRemoteTrackSourceNames.forEach(sourceName => {
+                receiverConstraints.constraints[sourceName] = { 'maxHeight': maxFrameHeightForVerticalFilmstrip };
             });
+        }
 
-            // Prioritize screenshare in tile view.
-            remoteScreenShares?.length && (receiverConstraints.selectedEndpoints = remoteScreenShares);
-
-        // Stage view.
-        } else {
-            if (!visibleRemoteParticipants?.size && !largeVideoParticipantId) {
-                return;
-            }
+        if (getCurrentLayout(state) === LAYOUTS.STAGE_FILMSTRIP_VIEW && activeParticipantsSources.length > 0) {
+            const onStageSources = [ ...activeParticipantsSources ];
 
-            if (visibleRemoteParticipants?.size > 0) {
-                const qualityLevel = getVideoQualityForResizableFilmstripThumbnails(state);
-                const stageParticipantsLevel = getVideoQualityForStageThumbnails(state);
+            activeParticipantsSources.forEach(sourceName => {
+                const isScreenSharing = remoteScreenShares.includes(sourceName);
+                const quality
+                    = isScreenSharing && preferredVideoQuality >= MAX_VIDEO_QUALITY
```

Comment:
What happens here when the preferredQuality is set to lower than 2160 and the send resolution on the SS stream being requested is higher than 2160 ? Does the bridge forward it anyways ?

Reply:
@bgrozev can explain this better.

But what I understood from him is that on VP9 the 3 layers are with the same resolution anyway and the constraints we set are mostly used to figure out what the user has selected as preferredVideoQuality... He was also saying it should be fine for VP8 too.

Reply:
I would think all the layers would have same resolution for low fps sharing since the client sends only 1 stream/spatial layer irrespective of the codec. Only when high fps sharing is enabled, all the streams/layer with different resolutions will be enabled. Nonetheless, it will be worthwhile to test all the scenarios, especially when the SS sender does both low and high fps sharing and with both VP8 and VP9 codec.

Reply:
When I tested it was looking good for me! 

Anyway as I said you may check with @bgrozev what actually happens on the bridge side.